Code/Framework/AzCore/AzCore/IO/GenericStreams.h

```diff
@@ -61,38 +62,47 @@ namespace AZ::IO
     /*
      * Wraps around a SystemFile
      */
-    class SystemFile;
     class SystemFileStream
         : public GenericStream
     {
     public:
-        SystemFileStream(SystemFile* file, bool isOwner, SizeType baseOffset = 0, SizeType fakeLen = static_cast<SizeType>(-1));
+        //! Creates a SystemFile stream with an underlying SystemFile which is not open
+        //! The Open() function can be used to Open the SystemFile
+        SystemFileStream();
+        //! Opens a SystemFile stream using the specific path and OpenMode
+        SystemFileStream(const char* path, OpenMode mode);
+        //! SystemFiles are only movable and only one instance is available at a time.
+        SystemFileStream(SystemFile&& file);
+        //! For this overload, the mode parameter
+        //! is for passing the mode flags used to open the supplied SystemFile
+        //! This purpose of the mmode parameter is only for use in the ReOpen() function
+        //! to determine how to reopen the SystemFile
+        SystemFileStream(SystemFile&& file, OpenMode mode);
         ~SystemFileStream() override;
 
         bool Open(const char* path, OpenMode mode);
 
         void Close() override;
 
         bool        IsOpen() const override;
-        bool        CanSeek() const override { return true; }
+        bool        CanSeek() const override;
         bool        CanRead() const override;
         bool        CanWrite() const override;
         void        Seek(OffsetType bytes, SeekMode mode) override;
         SizeType    Read(SizeType bytes, void* oBuffer) override;
         SizeType    Write(SizeType bytes, const void* iBuffer) override;
-        SizeType    GetCurPos() const override { return m_curPos; }
+        SizeType    GetCurPos() const override;
         SizeType    GetLength() const override;
         const char* GetFilename() const override;
-        OpenMode    GetModeFlags() const override { return m_mode; }
+        OpenMode    GetModeFlags() const override;
         bool        ReOpen() override;
 
+        //! Moves a SystemFile instance from an rvalue or xvalue SystemFileStream
+        SystemFile MoveSystemFile() &&;
```

Comment:
Is this function used anywhere?  I didn't see it.

What is this syntax called btw?

Reply:
This function is not used anywhere yet.

This syntax here is a [reference qualifier ](https://en.cppreference.com/w/cpp/language/member_functions#member_functions_with_ref-qualifier)

The purpose of this function is allow the use to move or "steal" the SystemFile from an expiring `SystemFileStream`
ex.
```c++
SystemFile WriteToTextFile()
{
    // Open up normal SystemFile and do work with it
    SystemFile systemFile("file.txt", SystemFile::OpenMode::SF_OPEN_WRITE_ONLY);
    // ...
    // Need to interact with API which uses GenericStream?
    // Give ownership of the SystemFile to a SystemFileStream
    SystemFileStream systemFileStream(AZStd::move(systemFile));
    // Use AZ::Utils::SaveObjectToStream to serialize class to SystemFile
   AZ::Utils::SaveObjectToStream(systemFileStream, ...);

   // Give ownership of the created SystemFile back to the "systemFile" variable
   systemFile = AZStd::move(systemFileStream).MoveSystemFile();

   return systemFile;
}
```

Reply:
Thanks!